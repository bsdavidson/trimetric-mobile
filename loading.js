import React, {Component} from "react";
import {View, Text, ActivityIndicator, StyleSheet} from "react-native";
import {getVehiclePoints, getStopPoints} from "./selectors";
import {connect} from "react-redux";

import {setLoadingStatusLoaded} from "./actions";

export class Loading extends Component {
  constructor(props) {
    super(props);
    this.state = {
      loading: true
    };

    this.loadingStatus = 0;
  }

  componentDidUpdate() {
    if (!this.state.loading) {
      return;
    }
    if (!this.props.totals) {
      return;
    }
    let thingsLoaded =
      this.props.stopPoints.features.length +
      (this.props.routeShapes && this.props.routeShapes.length);
    let totalThingsToLoad =
      this.props.totals.stops + this.props.totals.route_shapes;

    this.loadingStatus = Math.min(
      Math.floor(thingsLoaded / totalThingsToLoad * 100),
      100
    );

    if (this.props.stopPoints.features.length < this.props.totals.stops) {
      return;
    }

    if (
      !this.props.routeShapes ||
      this.props.routeShapes.length < this.props.totals.route_shapes
    ) {
      return;
    }

    if (!this.props.receivedVehicles) {
      return;
    }

    this.props.onLoadComplete();
    this.setState({
      loading: false
    });
  }

  render() {
    if (!this.props.connected) {
      return (
        <View style={styles.container}>
          <ActivityIndicator
            animating={this.state.loading}
            size="small"
            color="#FFFFFF"
          />
          <Text style={[styles.text]}>Connecting to server...</Text>
        </View>
      );
    }

    return (
      <View style={styles.container}>
        <ActivityIndicator
          animating={this.state.loading}
          size="small"
          color="#FFFFFF"
        />

        <Text style={[styles.text, {opacity: this.state.loading ? 1 : 0}]}>
          Loading GTFS data...
          {/* <Text style={{color: "#FFFFFF"}}>{this.loadingStatus}%</Text> */}
        </Text>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    alignItems: "flex-start",
    flex: 1,
    flexDirection: "row",
    justifyContent: "center"
  },
  text: {
    color: "#ecf0f1",
    marginLeft: 10
  }
});

function mapDispatchToProps(dispatch) {
  return {
    onLoadComplete: items => {
      dispatch(setLoadingStatusLoaded());
    }
  };
}

function mapStateToProps(state) {
  return {
    connected: state.connected,
    receivedVehicles: state.receivedVehicles,
    routeShapes: state.routeShapes,
    stopPoints: getStopPoints(state),
    totals: state.totals
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(Loading);
