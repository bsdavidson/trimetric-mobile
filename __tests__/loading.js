import "react-native";
import React from "react";
import renderer from "react-test-renderer";

import {Loading} from "../loading";

it("renders loading component", () => {
  const connecting = renderer.create(<Loading />).toJSON();
  expect(connecting).toMatchSnapshot("loading_connecting");

  const connected = renderer.create(<Loading connected={true} />).toJSON();
  expect(connected).toMatchSnapshot("loading_connected");
});
