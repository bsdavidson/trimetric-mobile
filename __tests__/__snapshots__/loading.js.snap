// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loading_connected 1`] = `
<View
  style={
    Object {
      "alignItems": "flex-start",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#FFFFFF"
    hidesWhenStopped={true}
    size="small"
  />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#ecf0f1",
          "marginLeft": 10,
        },
        Object {
          "opacity": 1,
        },
      ]
    }
  >
    Loading GTFS data...
  </Text>
</View>
`;

exports[`loading_connecting 1`] = `
<View
  style={
    Object {
      "alignItems": "flex-start",
      "flex": 1,
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#FFFFFF"
    hidesWhenStopped={true}
    size="small"
  />
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#ecf0f1",
          "marginLeft": 10,
        },
      ]
    }
  >
    Connecting to server...
  </Text>
</View>
`;
